### DISCUSSION TITLE
Running Ceedling for different configuration does not work in v1.0.0?

### DISCUSSION THREAD

**Discussion by kozmotronik** [ThrowTheSwitch/Ceedling#932] [Q&A]
Hello,
Hope you all are doing well.
Upon [this discussion](https://github.com/ThrowTheSwitch/Ceedling/discussions/845#discussion-6213791) we had before, I managed to create a release config file and release app code successfully using the `ceedling project:release release` command as per the [JuPrgn's answer](https://github.com/ThrowTheSwitch/Ceedling/discussions/845#discussioncomment-8514105).
Recently I just was upgraded the ceedling version to the following:
```
$ ceedling version
🌱 Welcome to Ceedling!

  Ceedling => 1.0.0-7711e2b
  ----------------------
  /var/lib/gems/3.0.0/gems/ceedling-1.0.0/
  
  Build Frameworks
  ----------------------
       CMock => 2.5.4
       Unity => 2.6.0
  CException => 1.3.4
```
After this upgrade I can't release my project anymore. The command I use and its output is:
```
$ ceedling project:release release
🚧 Loaded project configuration from working directory.
 > Using: /disk/Projeler/HimtecKontrol/Bellenim/hkbmv2/hkbmv2/project.yml
 > Working directory: /disk/Projeler/HimtecKontrol/Bellenim/hkbmv2/hkbmv2

Ceedling set up completed in 106 milliseconds

🧨 EXCEPTION: Don't know how to build task 'project:release' (See the list of available tasks with `rake --tasks`)
🌱 Ceedling could not complete operations because of errors
```
This command was working in the v0.32.0 release. So doesn't ceedling recognize the `project` task any more? Should I downgrade its version?

**Comment by mvandervoord** [comment]
The old `project:ceedling` functionality (as well as user files, config files, etc) have all been updated with the new `mixin` handling. You can read about it here:

https://github.com/ThrowTheSwitch/Ceedling/blob/test/ceedling_0_32_rc/docs/CeedlingPacket.md#mixins-example-plus-merging-rules

In this case, it's as simple as using the following (assuming all other breaking changes have been updated):

```
ceedling --project=release.yml release
```

**Reply by JuPrgn** [reply]
@kozmotronik I will also switch to 1.0.0 on next weeks for a new project. There are some really interesting changes but probably some work to switch to it please share some details

**Reply by kozmotronik** [reply]
Thank you for a lightning fast response @mvandervoord!

@JuPrgn When I have a look at the `mixin` topic, I think I will need a couple of hours to digest the required information to properly update or reconfigure the relevant files.

For now, I've just tried to execute the command with the new offered syntax but no avail as expected 😅😆. This is what I get as a response from ceedling:
```
$ ceedling clobber --project=release-ch1.yml release
🚧 Loaded project configuration from command line argument.
 > Using: /disk/Projeler/HimtecKontrol/Bellenim/hkbmv2/hkbmv2/release-ch1.yml
 > Working directory: /disk/Projeler/HimtecKontrol/Bellenim/hkbmv2/hkbmv2
🪲 ERROR: Required config file entry :paths does not exist.
🪲 ERROR: Required config file entry :project ↳ :build_root does not exist.
🪲 ERROR: Required config file entry :paths ↳ :test does not exist.
🪲 ERROR: Required config file entry :paths ↳ :source does not exist.

🧨 EXCEPTION: Ceedling configuration failed validation
```
So I'll let you know how when I do everything properly as per the documentation and get it working.

**Reply by kozmotronik** [reply]
I've reached a mile stone. I just tried to fix the errors specified in ceedling's output. Here is what I've done in brief steps...

## 1. Add the `paths` to the release-ch1.yml
I copied the `paths` section from the ***project.yml*** and paste it into the ***release-ch1.yml*** as is. This cleared the three of the four errors that ceedling was complaining.

## 2. Add the `project` section to the release.yml
I also copied the `project` whose subsection is `build_root` as is. After that all the previous errors are cleared.

## 3. Update XC8 and also its system path
I updated XC8 to its v2.50 version and also its system path.

It is almost done IMO, but still missing something. Here is the output after getting rid of the previous errors:
```
$ ceedling clobber --project=release-ch1.yml release
🚧 Loaded project configuration from command line argument.
 > Using: /disk/Projeler/HimtecKontrol/Bellenim/hkbmv2/hkbmv2/release-ch1.yml
 > Working directory: /disk/Projeler/HimtecKontrol/Bellenim/hkbmv2/hkbmv2

Ceedling set up completed in 95 milliseconds

Clobbering all generated files...
(For large projects, this task may take a long time to complete)

🧨 EXCEPTION: Don't know how to build task 'release' (See the list of available tasks with `rake --tasks`)
🌱 Ceedling could not complete operations because of errors
```
As far as the **mixins** concept, the idea behind looks virtually identical to including header files in C. The only difference is, the syntax of inclusion in yaml files.
As far as I understand you can have a bunch of yaml configuration files that each has a specific configuration delta for different types of builds (called mixin files). Then you declare the directory and the names of these mixin files in the base.yml file. This way ceedling merge those files upon a compilation.
I will keep reading again and again until I digest it fully 😄

**Reply by kozmotronik** [reply]
Hi @JuPrgn,

I ended up creating a whole separate configuration file in order to get it compiling for release. Before the v1.0.0 ceedling was using the delta of the configuration files; *project.yml* and *release.yml* respectively. The *project.yml* is configured with native GCC compiler for testing while the *release.yml* is configured for release build with platform specific XC8 compiler.

In the last update as you can see in the output, there was no `:release_build: TRUE` statement in *release.yml*, that's why was the error. Now that I've copied the `:project:` and the `:paths:` properties with all their respective values from the *project.yml*, then issuing the command with new switch syntax as purposed: `ceedling --project=release.yml release`, the compilation is successful now.

## In summary
You just have to copy the `:project:` and the `:paths:` properties with all their respective values from the *project.yml* or whatever you name your test configuration. Or Better, if you are using a single configuration file for your project both for testing and releasing, you don't even have to change anything at all. Just issue the command with its new syntax and you are good to go.

**Reply by JuPrgn** [reply]
Hi @kozmotronik !
Great thanks for sharing! This looks pretty simple after all, I read a bit about mixins it looked very versatile but quite complex. 
Nice to have a Microchip buddy here 😄
